metricsaggregator/src/main/java/io/dockstore/metricsaggregator/MetricsAggregatorS3Client.java

```diff
@@ -125,33 +129,87 @@ public void aggregateMetrics(ExtendedGa4GhApi extendedGa4GhApi) {
+        // getMetricsData uses the S3 ListObjectsV2Request which returns objects in alphabetical order.
+        // Since the file names are the time of submission in milliseconds, metricsDataList is sorted from oldest file name to newest file name
         List<MetricsData> metricsDataList = metricsDataS3Client.getMetricsData(toolId, versionName, Partner.valueOf(platform));
-        List<RunExecution> runExecutionsFromAllSubmissions = new ArrayList<>();
-        List<TaskExecutions> taskExecutionsFromAllSubmissions = new ArrayList<>();
-        List<ValidationExecution> validationExecutionsFromAllSubmissions = new ArrayList<>();
-        List<Metrics> aggregatedExecutionsFromAllSubmissions = new ArrayList<>();
+        Map<String, RunExecution> executionIdToWorkflowExecutionMap = new HashMap<>();
+        Map<String, TaskExecutions> executionIdToTaskExecutionsMap = new HashMap<>();
+        Map<String, ValidationExecution> executionIdToValidationExecutionMap = new HashMap<>();
+        Map<String, AggregatedExecution> executionIdToAggregatedExecutionMap = new HashMap<>();
 
         for (MetricsData metricsData : metricsDataList) {
             String fileContent = metricsDataS3Client.getMetricsDataFileContent(metricsData.toolId(), metricsData.toolVersionName(),
                     metricsData.platform(), metricsData.fileName());
-            ExecutionsRequestBody executionsFromOneSubmission = GSON.fromJson(fileContent, ExecutionsRequestBody.class);
-            runExecutionsFromAllSubmissions.addAll(executionsFromOneSubmission.getRunExecutions());
-            taskExecutionsFromAllSubmissions.addAll(executionsFromOneSubmission.getTaskExecutions());
-            validationExecutionsFromAllSubmissions.addAll(executionsFromOneSubmission.getValidationExecutions());
-            aggregatedExecutionsFromAllSubmissions.addAll(executionsFromOneSubmission.getAggregatedExecutions());
+
+            ExecutionsRequestBody executionsFromOneSubmission;
+            try {
+                executionsFromOneSubmission = GSON.fromJson(fileContent, ExecutionsRequestBody.class);
+            } catch (JsonSyntaxException e) {
+                LOG.error("Could not read execution(s) from S3 key {}, ignoring file", metricsData.s3Key());
+                continue;
+            }
+
+            // For each execution, put it in a map so that there are no executions with duplicate execution IDs.
+            // The latest execution put in the map is the newest one based on the principal that S3 lists objects in alphabetical order,
+            // which is returned in an ordered list via getMetricsData.
+            // Note: executions that were submitted to S3 prior to the existence of execution IDs don't have an execution ID.
+            // For the purposes of aggregation, generate one so that the execution is considered unique.
```

Comment:
We could also wipe the old executions and submit anew with execution IDs.

Reply:
Hmm, true, everything in the metrics S3 bucket is submitted by us and it would allow us to get rid of this special case. If we're okay with that, sounds like we can do the following steps:

1. Wipe the entire metrics bucket (everything in it was submitted by us)
2. Re-submit the DNAstack validation metrics using the [submit-validation-data](https://github.com/dockstore/dockstore-support/blob/develop/metricsaggregator/README.md#submit-validation-data) command
3. Re-upload the tool tester AGC executions using the [upload-results](https://github.com/dockstore/dockstore-support/tree/develop/tooltester/metrics-created-during-testing) command, which uploads the executions in this [directory](https://github.com/dockstore/dockstore-support/tree/develop/tooltester/metrics-created-during-testing). There's not many so a quick fix is to manually modify those files to include execution IDs.
4. Ingest Terra metrics
5. Aggregate metrics

@denis-yuen if we're okay with that, I'll take out that PR change that I commented on and add release notes steps to do the above. Thoughts?

Reply:
I updated the PR so that we can wipe the metrics bucket and re-ingest the DNAstack validation metrics and tool tester AGC executions.
- For the DNAstack validation metrics, I modified the `metricsaggregator/scripts/format-dnastack-validation-data.sh` script so that it takes a dateExecuted argument that is used by the executions submitted to Dockstore. This way, we can set the dateExecuted to the date of the first time we submitted the metrics
- For the tooltester AGC executions, I added a unique UUID as the execution ID for each execution, and I had to add a dateExecuted field because that's required by our schema now. Luckily, the date executed is the file name, so I just converted those epoch milliseconds to ISO date format

I will add the steps mentioned in my previous comment to the release checklist. @denis-yuen  re-requesting your review